util/fipstools/test_fips.c

```diff
@@ -140,8 +149,128 @@ int main(int argc, char **argv) {
+  OPENSSL_memset(aes_iv, 0, sizeof(aes_iv));
+  if (AES_set_decrypt_key(kAESKey, 8 * sizeof(kAESKey), &aes_key) != 0) {
+    printf("AES decrypt failed\n");
+    goto err;
+  }
+  printf("About to AES-CCM open ");
+  hexdump(output, out_len);
+  if (!EVP_AEAD_CTX_open(&aead_ctx, output, &out_len, sizeof(output), nonce,
+                        EVP_AEAD_nonce_length(EVP_aead_aes_128_ccm_bluetooth()),
+                        output, out_len, NULL, 0))
+  {
+    fprintf(stderr, "EVP_AEAD_CTX_open for AES-128-CCM failed.\n");
+    goto err;
+  }
+  printf("  got ");
+  hexdump(output, out_len);
+  
+  OPENSSL_cleanse(&aes_key, sizeof(aes_key));
+  EVP_AEAD_CTX_zero(&aead_ctx);
+
+  /* AES-ECB */
+  /* AES-ECB Encryption */
+  OPENSSL_memset(aes_iv, 0, sizeof(aes_iv));
+  if (AES_set_encrypt_key(kAESKey, 8 * sizeof(kAESKey), &aes_key) != 0)
+  {
+    fprintf(stderr, "AES_set_encrypt_key failed.\n");
+    goto err;
+  }
+  printf("About to AES-ECB encrypt ");
+  hexdump(output, out_len);
+  for (size_t j = 0; j < sizeof(kPlaintext) / 16; j++)
+  {
+    AES_ecb_encrypt(&kPlaintext[j * 128], & output[j * 128], &aes_key, AES_ENCRYPT);
+  }
+  printf("  got ");
+  hexdump(output, out_len);
+
+  /* AES-ECB Decryption */
+  OPENSSL_memset(aes_iv, 0, sizeof(aes_iv));
+  if (AES_set_decrypt_key(kAESKey, 8 * sizeof(kAESKey), &aes_key) != 0) {
+    printf("AES decrypt failed\n");
+    goto err;
+  }
+  printf("About to AES-ECB decrypt ");
+  hexdump(output, out_len);
+  for (size_t j = 0; j < out_len / 16; j++)
+  {
+    AES_ecb_encrypt(&output[j * 128], & output[j * 128], &aes_key, AES_DECRYPT);
+  }
+  printf("  got ");
+  hexdump(output, out_len);
+
+  OPENSSL_cleanse(&aes_key, sizeof(aes_key));
+
+  unsigned int num = 0;
+  uint8_t ecount_buf[128];
+
+  /* AES-CTR */
+  /* AES-CTR Encryption */
+  OPENSSL_memset(aes_iv, 0, sizeof(aes_iv));
+  if (AES_set_encrypt_key(kAESKey, 8 * sizeof(kAESKey), &aes_key) != 0)
+  {
+    fprintf(stderr, "AES_set_encrypt_key failed.\n");
+    goto err;
+  }
+  printf("About to AES-CTR Encrypt ");
+  hexdump(output, out_len);
+  AES_ctr128_encrypt(kPlaintext, output, sizeof(kPlaintext), &aes_key, aes_iv, ecount_buf, &num);
+  printf("  got ");
+  hexdump(output, out_len);
+
+
+  /* AES-CTR Decryption */
+  OPENSSL_memset(aes_iv, 0, sizeof(aes_iv));
+  printf("About to AES-CTR Decrypt ");
+  hexdump(output, out_len);
+  AES_ctr128_encrypt(output, output, out_len, &aes_key, aes_iv, ecount_buf, &num);
+  printf("  got ");
+  hexdump(output, out_len);
+
+  OPENSSL_cleanse(&aes_key, sizeof(aes_key));
+
+  /* AES-KW Wrap */
+  if (AES_set_encrypt_key(kAESKey, 8 * sizeof(kAESKey), &aes_key) != 0)
+  {
+    fprintf(stderr, "AES_set_encrypt_key failed.\n");
+    goto err;
+  }
+  printf("About to AES-KW Wrap ");
+  hexdump(output, out_len);
+  AES_wrap_key(&aes_key, NULL, output, kPlaintext, sizeof(kPlaintext));
+  printf("  got ");
+  hexdump(output, out_len);
+
+  OPENSSL_cleanse(&aes_key, sizeof(aes_key));
+
```

Comment:
Do we need to do an `AES_unwrap_key` here?

Reply:
ATSEC doesn't _need_ it per se but it can't hurt to put it in there for consistency lol